web/modules/custom/collection/src/Plugin/Filter/Glossary.php

```diff
@@ -0,0 +1,248 @@
+          continue;
+        }
+
+        // @todo Consider replacing the link with a JavaScript tooltip.
+        $link = $document->createElement('a', $text_part);
+        $link->setAttribute('href', $replacements[$text_part]['url']);
+        $link->setAttribute('class', 'glossary-term');
+        $link->setAttribute('title', $replacements[$text_part]['summary']);
+        $parent_node->insertBefore($link, $text_node);
+      }
+      $parent_node->removeChild($text_node);
+    }
+
+    return $result
+      ->setProcessedText($document->saveHTML())
+      ->addCacheableDependency($this->cacheMetadata);
+  }
+
+  /**
+   * Finds-out if a piece of text is inside a HTML link.
+   *
+   * @param \DOMText $text_node
+   *   The text node to be checked.
+   *
+   * @return bool
+   *   If the given text node is inside a HTML link.
+   */
+  protected function isLinkText(\DOMText $text_node): bool {
+    $node = $text_node;
+    do {
+      $node = $node->parentNode;
+      if (!$node->parentNode) {
+        return FALSE;
+      }
+    } while ($node->tagName !== 'a');
+    return TRUE;
+  }
+
+  /**
+   * Builds and returns a replacements map.
+   *
+   * @return array
+   *   Associative array keyed by the glossary term or abbreviation. The values
+   *   are arrays with two keys:
+   *   - url: The glossary term URL.
+   *   - summary: A summary to be used as tooltip.
+   */
+  protected function getReplacementsMap(): array {
+    // Make sure the filter cache invalidates when a new glossary term is added
+    // in this collection.
+    $this->cacheMetadata = (new CacheableMetadata())
+      ->addCacheTags(
+        Cache::buildTags('og-group-content', $this->collection->getCacheTagsToInvalidate())
+      );
+
+    $map = [];
+    $node_storage = $this->entityTypeManager->getStorage('node');
+    $nids = $node_storage->getQuery()
+      ->condition('type', 'glossary')
+      ->condition(OgGroupAudienceHelperInterface::DEFAULT_FIELD, $this->collection->id())
+      ->condition('status', TRUE)
+      ->execute();
+
+    if ($nids) {
+      /** @var \Drupal\node\NodeInterface $glossary */
+      foreach ($node_storage->loadMultiple($nids) as $glossary) {
+        $link = [
+          'url' => $glossary->toUrl()->toString(),
+          'summary' => $this->getSummaryOrTrimmedDefinition($glossary),
+        ];
+
+        $map[$glossary->label()] = $link;
+        // Link also the abbreviation, if any.
+        if (!$glossary->get('field_glossary_abbreviation')->isEmpty()) {
+          $map[$glossary->get('field_glossary_abbreviation')->value] = $link;
+        }
+
+        // When this glossary node is changing, invalidate the filter cache.
+        $this->cacheMetadata->addCacheableDependency($glossary);
+      }
+    }
+
+    return $map;
+  }
+
+  /**
+   * Stores the collection context.
+   */
+  protected function setCollection(): void {
+    if ($this->collection = $this->ogContext->getGroup()) {
```

Comment:
This is not 100% reliable since these texts can also be shown outside of a collection (e.g. in the search results which have no group context). With the render cache in play this could lead to unexpected results - for example if some content is rendered as a tile and is first displayed in the search results it will be cached without having its glossary terms replaced. Then when the cached render array is later used inside the collection context the terms will still be missing.

I am going to let this pass though, for the following reasons:

- I could not replicate this in local testing. Search results and other places where tiles are shown (e.g. Keep Up To Date) correctly show the content without glossary terms. Possibly there is a render context that causes variations to be cached.
- This is a [known problem](https://www.drupal.org/project/drupal/issues/226963) in Drupal for a very long time, and will be incredibly hard to fix, way out of scope for this PR.
- I created a followup to look into this in some more depth, we can skip it for now: ISAICP-6116

Reply:
OK, let's see the feedback. The customer is doesn't have yet a 100% opinion on the business logic.